### PR TITLE
Atomic bridge

### PR DESCRIPTION
**Pull request by andyjsbell**
## Description

The atomic bridge as specified https://github.com/movementlabsxyz/rfcs/blob/main/0040-atomic-bridge/rfc-0040-atomic-bridge.md ported from https://github.com/movementlabsxyz/movement to bring the native bridge within the framework creating a native bridge from ERC20 Move to Move L2.  

Inspiration was taken from the original work done on https://github.com/movementlabsxyz/movement.  Kudos to @andygolay and @0xmovses for this.

With moving to the framework a few things have been able to be normalised to in most part reduce code repetition.  The bridge on genesis acquires the Mint and Burn capabilities of the native coin with which it bridges to and from L1. 

The role of bridge operator, or signer for the relayer, is configurable via governance.  Time locks for both counterparty and initiator are also configurable via governance.

Tests can be run for the framework 

`aptos move test --package-dir aptos-move/framework/aptos-framework`

There is a start in formally verifying the bridge which still needs more work.

`aptos move prove --package-dir aptos-move/framework/aptos-framework --filter atomic_bridge`

update:
The feature `ATOMIC_BRIDGE` has been added to `0x1::features` which is disabled at genesis and activated via governance.

update:
We now piggy back `transaction_fee.move` for caps for native bridge instead of at genesis

## Type of Change
- [x] New feature
- [ ] Bug fix
- [x] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [x] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/atomic_bridge.move
Comment (andygolay) [inline aptos-move/framework/aptos-framework/sources/atomic_bridge.move:1072]: Since we're using timestamp, `block::initialize_for_test` is probably not needed. Tests pass without it.

**Review by ganymedio — CHANGES_REQUESTED** [review]
- Move unit tests and Move Prover verification pass. Still looking into the details of both to ensure soundness. It would be good to have Research and other engineers vet as well.

- Reviewing against auditor feedback:

  - Time lock verification for `atomic_bridge_counterparty::complete_bridge_transfer` [(HAL-11)](https://github.com/movementlabsxyz/movement/issues/568): 

  - `complete_bridge_transfer` calls `bridge_store::complete_transfer` that calls `complete_details` which checks the validity and correctness of the hash lock as well as that the state is pending. But I don't see it checking the time lock with something along the lines of:

```
assert!(timestamp::now_seconds() <= bridge_transfer.time_lock, ETIMELOCK_EXPIRED);
```
  - Recommendation: add this check into `complete_details` with other completion checks.

File: aptos-move/framework/aptos-framework/sources/atomic_bridge.move
Comment (andygolay) [inline aptos-move/framework/aptos-framework/sources/atomic_bridge.move:731]: Add time lock verification here?

**Comment by andyjsbell** [comment]
Test and coverage, I believe @franck44 will be looking further into the specification side of thing.  I will also look at the spec again as this was my first bash at it. 

@andygolay you may find this interesting for the e2e work you have been exploring which is part of this PR.
https://github.com/movementlabsxyz/aptos-core/blob/d76b853b019491ff8cd3b3cf85fbc96963051aa3/aptos-move/e2e-move-tests/src/tests/bridge.rs#L35

Some coverage results.  

```
Module 0000000000000000000000000000000000000000000000000000000000000001::ethereum
>>> % Module coverage: 100.00
Module 0000000000000000000000000000000000000000000000000000000000000001::atomic_bridge_store
>>> % Module coverage: 92.40
Module 0000000000000000000000000000000000000000000000000000000000000001::atomic_bridge_configuration
>>> % Module coverage: 95.45
Module 0000000000000000000000000000000000000000000000000000000000000001::atomic_bridge
>>> % Module coverage: 81.82
Module 0000000000000000000000000000000000000000000000000000000000000001::atomic_bridge_counterparty
>>> % Module coverage: 100.00
Module 0000000000000000000000000000000000000000000000000000000000000001::atomic_bridge_initiator
>>> % Module coverage: 100.00

```

Would be great to get CI up and running as the e2e tests were broken, some time soon :)

**Review by ganymedio — DISMISSED** [review]

**Comment by andygolay** [comment]
LGTM for testnet iteration.

**Comment by andygolay** [comment]
Added issue here for CI implementation https://github.com/movementlabsxyz/aptos-core/issues/73

**Comment by andyjsbell** [comment]
> LGTM for testnet iteration. Looking into the E2E tests more deeply now, thanks for highlighting those.
No worries.  I think we would want to see this on devnet first though.  I have purposely added [DO NOT MERGE] to the PR for the time being as I think its best to feature flag this before merging

**Comment by andygolay** [comment]
Yep, devnet then testnet + feature flag makes good sense.

**Review by 0xmovses — APPROVED** [review]
LGTM

**Review by ganymedio — APPROVED** [review]
LGTM

**Pull request merged**
